rstudio/rstudio
# Show memory usage details in Environment Pane

### Intent

This change adds a small feature that makes it possible to see your current memory consumption in the Environment Pane. This is a feature frequently requested by users that run R in memory constrained environments, especially with RStudio Server. 

Addresses https://github.com/rstudio/rstudio/issues/4033.

### Approach

RStudio's Environment Pane now includes the following memory usage widget:

![image](https://user-images.githubusercontent.com/470418/106518278-7d954400-648e-11eb-85f1-28959a6937c0.png)

This small bit of UI is placed here so that the buttons follow the workflow "import data into memory", "data currently in memory", "remove data from memory". It conveys two important values:

- The total amount of used/available memory (pie chart)
- The total memory used by the R session (MiB value)

The pie chart also has color coding for usage thresholds; it turns yellow and orange when usage approaches the limit, and bright red when very little memory is remaining. 

![image](https://user-images.githubusercontent.com/470418/106518562-eda3ca00-648e-11eb-82ae-deaa316c9dfa.png)

Hovering over the values provides a little more detail. There's also a new Memory Usage Report which breaks down the amount of memory currently used by R (obtained from a `gc()` call), the IDE, and the system itself.

![image](https://user-images.githubusercontent.com/470418/106518704-2479e000-648f-11eb-9438-eda3b15f0f18.png)


While this feature is primarily of interest to RStudio Server users, it works on all platforms. This is done by using platform-specific implementations of a few functions that measure memory statistics. Backends have been implemented for Windows, macOS, and Linux (both `/proc/meminfo` and `cgroups` are supported).

It could be a little annoying to have a number in the Environment Pane that changes constantly as you work, so it's possible to turn this feature off if you don't like it. (We might consider turning it off by default on RStudio Desktop where memory is rarely top of mind.) This is done via a preference you can toggle from the menu.

![image](https://user-images.githubusercontent.com/470418/106518875-57bc6f00-648f-11eb-9cfc-2756fb8a846e.png)

Finally, memory usage is updated frequently: any time you enter an R command, and in the background while the session runs, every 10 seconds by default. This allows the memory usage indicator to show changes in usage that result from allocations in background jobs and other system processes. There is a new advanced option (accessible only via the Palette) which can be used to adjust the interval at which statistics are refreshed.

![image](https://user-images.githubusercontent.com/470418/106518766-39567380-648f-11eb-8015-403c892320b2.png)

This can be set to 0 to disable periodic refresh entirely.

### Automated Tests

This change includes unit tests that test both basic resource usage metrics and the memory report RPC.

### QA Notes

Lots of cross-platform testing is needed here since measuring memory usage is very OS-specific. On Linux in particular, the primary motivator for this change is for Kubernetes/Slurm environments, where memory usage is often constrained with cgroups.

This change is almost entirely additive; it doesn't touch much existing code, so risk to other areas is minimal. Probably the most relevant change is that the server is now trickling memory usage updates every 10 seconds to the client; formerly the client would often go a long time (minutes or longer) without updates from the server. Make sure things like session suspend and idle timeouts are still working with the increased traffic.

### Checklist
